### PR TITLE
fix: alias group_by exprs in single_distinct_to_groupby optimizer

### LINKED ISSUE #2994: single_distinct_to_groupby panic when group by expr is a binaryExpr
**Describe the bug**

single_distinct_to_groupby panic when group by expr is a binaryExpr, with error message like this:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SchemaError(FieldNotFound { qualifier: Some("users"), name: "id", valid_fields: Some(["users.id + Int64(1)", "alias1"]) })', datafusion/optimizer/src/single_distinct_to_groupby.rs:117:72
```

**To Reproduce**
Based on https://github.com/apache/arrow-datafusion/blob/e47c4eb1ac8a5c3c07a0ed2bad74ecd6509bae7b/datafusion-examples/examples/memtable.rs#L39,

Change sql to following will reproduce this error.
```sql
SELECT id+1 as id, count(distinct(bank_account)) 
FROM users
group by id+1
order by id
```

**Expected behavior**
No error

**Additional context**

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

# Which issue does this PR close?

Closes #2994.

 # Rationale for this change

The original issue is https://github.com/CeresDB/ceresdb/issues/59, and a minimal reproducer can be found in #2994.

This is trying to fix the `single_distinct_to_groupby` optimizer cannot handle a `GROUP BY` with an expression, like `GROUP BY c+1`. To fix this, we need to alias the "original" GROUP BY in the subquery and refer to that aliased column in outer. E.g.:
```sql
Before optimize
SELECT id+1, count(distinct(bank_account))
FROM users
GROUP BY id+1;
```
After optimize
```sql
SELECT group_by_alias1 as "id+1", count(bank_account)
FROM (
  SELECT id+1 as group_by_alias1, bank_account
  FROM users 
  GROUP BY id+1, bank_account
)
GROUP BY group_by_alias1;
```
# What changes are included in this PR?

- alias `group_by` exprs
- rename some symbols and add comment
- a test to cover this case `group_by_with_expr`

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/optimizer/src/single_distinct_to_groupby.rs [hunk 1]
```diff
@@ -86,64 +101,66 @@ fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
...
-
-                let grouped_schema = DFSchema::new_with_metadata(
-                    all_field,
+                let inner_schema = DFSchema::new_with_metadata(
+                    inner_fields,
                     input.schema().metadata().clone(),
                 )
                 .unwrap();
-                let grouped_agg = LogicalPlan::Aggregate(Aggregate {
+                let grouped_aggr = LogicalPlan::Aggregate(Aggregate {
```
Comment (andygrove) [inline datafusion/optimizer/src/single_distinct_to_groupby.rs:118]: Please use `Aggregate::try_new` here, once https://github.com/apache/arrow-datafusion/pull/3286 is merged

**Comment by andygrove** [comment]
test failure caused by https://github.com/apache/arrow-datafusion/issues/3327

File: datafusion/optimizer/src/single_distinct_to_groupby.rs (hunk 1 above)
Comment (waynexia) [reply datafusion/optimizer/src/single_distinct_to_groupby.rs]: have removed all `unwrap`s

**Review by alamb — APPROVED** [review]
This looks good to me -- thanks @waynexia

**Comment by alamb** [comment]
Sorry for the delay in reviews -- I was out for a week and the DataFusion has been very busy

**Comment by alamb** [comment]
Screen Shot 2022-09-06 at 10 50 13 AM

I took the liberty of merging this branch with `master` to resolve merge conflicts. I'll plan to merge it when CI passes

**Comment by waynexia** [comment]
>I took the liberty of merging this branch with master to resolve merge conflicts.

Thanks! I've reviewed 94c014e and it looks fine 👍

**Comment by alamb** [comment]
Thanks again @waynexia

**Pull request merged**